app/bin/tools/merge_users.dart

```diff
@@ -0,0 +1,143 @@
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:args/args.dart';
+import 'package:gcloud/db.dart';
+
+import 'package:pub_dartlang_org/account/models.dart';
+import 'package:pub_dartlang_org/frontend/models.dart';
+import 'package:pub_dartlang_org/frontend/service_utils.dart';
+import 'package:pub_dartlang_org/history/models.dart';
+
+final _argParser = ArgParser()
+  ..addOption('source-user-id',
+      help: 'The userId of the User that will be removed.')
+  ..addOption('target-user-id',
+      help:
+          'The userId of the User that will take over the entries of the removed User.')
+  ..addFlag('help', abbr: 'h', defaultsTo: false, help: 'Show help.');
+
+Future main(List<String> args) async {
+  final argv = _argParser.parse(args);
+  if (argv['help'] as bool == true) {
+    print(
+        'Usage: dart merge_users.dart --source-user-id [uuid] --target-user-id [uuid]');
+    print(
+        'Merges one user into another (moving the ownership over to a new one).');
+    print(_argParser.usage);
+    return;
+  }
+
+  final sourceUserId = argv['source-user-id'] as String;
+  final targetUserId = argv['target-user-id'] as String;
+  if (sourceUserId == null ||
+      sourceUserId.isEmpty ||
+      targetUserId == null ||
+      targetUserId.isEmpty) {
+    print('Both user-id parameter must be set.');
+    print(_argParser.usage);
+    exit(-1);
+  }
+  if (sourceUserId == targetUserId) {
+    print('Use different user-ids.');
+    exit(-1);
+  }
+
+  await withProdServices(() async {
+    final sourceUser = (await dbService
+            .lookup<User>([dbService.emptyKey.append(User, id: sourceUserId)]))
+        .single;
+    final targetUser = (await dbService
+            .lookup<User>([dbService.emptyKey.append(User, id: targetUserId)]))
+        .single;
+    if (targetUser == null) {
+      print('Target user does not exists.');
+      exit(-1);
+    }
+    if (targetUser.oauthUserId == null) {
+      print('Target user does not have oauth user id.');
+      exit(-1);
+    }
+    print(
+        'Transferring ownership from ${sourceUser.email} to ${targetUser.email}.');
+    await Future.delayed(Duration(seconds: 5));
+
+    print('Querying Packages...');
+    final pkgQuery = dbService.query<Package>()
+      ..filter('uploaders =', sourceUserId);
+    final packages = await pkgQuery.run().toList();
+
+    print('Querying PackageVersions...');
+    final pvQuery = dbService.query<PackageVersion>()
+      ..filter('uploader =', sourceUserId);
+    final versions = await pvQuery.run().toList();
+
+    print('Found: ${packages.length} packages and '
+        '${versions.length} versions.');
+
+    final historyPackages = packages.map((p) => p.name).toSet();
+    versions.forEach((pv) => historyPackages.add(pv.package));
+    for (String package in historyPackages) {
+      print('Querying History for package $package...');
+      final query = dbService.query<History>()
+        ..filter('packageName =', package);
```

Comment:
This won't work for packages that the user used to be an uploader for? and where the user didn't upload any versions, but only added/removed other uploaders.

Reply:
We probably need to do a full table scan...

Reply:
updated